src/v/cloud_topics/level_one/compaction/log_sampler.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_topics/level_one/compaction/meta.h"
+#include "cloud_topics/level_one/metastore/metastore.h"
+#include "cluster/metadata_cache.h"
+#include "cluster/types.h"
+
+namespace cloud_topics::l1 {
+
+// A wrapper to provide easy mocking and break dependency on a multitude of
+// cluster objects within the `log_sampler`.
+class topic_cfg_provider {
+public:
+    virtual ~topic_cfg_provider() noexcept = default;
+
+    virtual std::optional<
+      std::reference_wrapper<const cluster::topic_configuration>>
+      get_topic_cfg(model::topic_namespace_view) const = 0;
+};
+
+// Default topic_cfg_provider, which is the `cluster::metadata_cache`.
+class topic_cfg_provider_impl : public topic_cfg_provider {
+public:
+    topic_cfg_provider_impl(cluster::metadata_cache*);
+
+    std::optional<std::reference_wrapper<const cluster::topic_configuration>>
+      get_topic_cfg(model::topic_namespace_view) const final;
+
+private:
+    cluster::metadata_cache* _metadata_cache;
+};
+
+// Responsible for issuing `get_compaction_info()` requests to the `metastore`
+// when attempting to schedule a round of compactions.
+class log_sampler {
+public:
+    log_sampler(metastore*, std::unique_ptr<topic_cfg_provider>);
+
+    // Populates `info_and_ts` within `log_compaction_meta`s from the provided
+    // `log_list_t` by sampling each log's compaction info from the metastore.
+    // It is not guaranteed that every log present in `log_list_t` will have its
+    // `info_and_ts` set e.g. due to concurrent removal or metastore errors.
+    // If a log already has `info_and_ts` set, it will not be sampled again
+    // until an interval has elapsed and the current `info_and_ts` is determined
+    // stale. Additionally, logs that have an inflight compaction in process do
+    // not need to be sampled. Logs that are sampled and deemed eligible for
+    // compaction will also have their `lw_shared_ptr` copied into the
+    // `log_compaction_queue` for future compaction.
+    ss::future<>
+    sample_logs(logs_type_t&, log_list_t&, log_compaction_queue&) const;
```

Comment:
nit: given how similar the two types are, I think these arguments could use some descriptive names

Also I think it'd be helpful to give cached_metadata a more descriptive name -- it seems like that's the main out-parameter, but that isn't quite clear from its name